# Workbook

Use this notebook to complete the exercises throughout the workshop.

#### Table of Contents
- [Section 1 &ndash; Getting Started with Pandas](#Section-1)
- [Section 2 &ndash; Data Wrangling](#Section-2)
- [Section 3 &ndash; Data Visualization](#Section-3)

--- 

### Section 1

#### Exercise 1.1
##### Create a DataFrame by reading in the `2019_Yellow_Taxi_Trip_Data.csv` file. Examine the first 5 rows.

In [21]:
import pandas as pd
taxi = pd.read_csv('../data/2019_Yellow_Taxi_Trip_Data.csv')
taxi.head()

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
0         2  2019-10-23T16:39:42.000  2019-10-23T17:14:10.000   
1         1  2019-10-23T16:32:08.000  2019-10-23T16:45:26.000   
2         2  2019-10-23T16:08:44.000  2019-10-23T16:21:11.000   
3         2  2019-10-23T16:22:44.000  2019-10-23T16:43:26.000   
4         2  2019-10-23T16:45:11.000  2019-10-23T16:58:49.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
0                1           7.93           1                  N   
1                1           2.00           1                  N   
2                1           1.36           1                  N   
3                1           1.00           1                  N   
4                1           1.96           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
0           138           170             1         29.5    1.0      0.5   
1            11            26             1         10.5    1.0      0.5   
2           163           162             1          9.5    1.0      0.5   
3           170           163             1         13.0    1.0      0.5   
4           163           236             1         10.5    1.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        7.98          6.12                    0.3         47.90   
1        0.00          0.00                    0.3         12.30   
2        2.00          0.00                    0.3         15.80   
3        4.32          0.00                    0.3         21.62   
4        0.50          0.00                    0.3         15.30   

   congestion_surcharge  
0                   2.5  
1                   0.0  
2                   2.5  
3                   2.5  
4                   2.5

In [22]:
taxi.columns

Index(['vendorid', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
       'passenger_count', 'trip_distance', 'ratecodeid', 'store_and_fwd_flag',
       'pulocationid', 'dolocationid', 'payment_type', 'fare_amount', 'extra',
       'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge',
       'total_amount', 'congestion_surcharge'],
      dtype='object')

In [23]:
taxi.index

RangeIndex(start=0, stop=10000, step=1)

In [24]:
taxi.tail()

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
9995         1  2019-10-23T17:39:59.000  2019-10-23T17:49:26.000   
9996         1  2019-10-23T17:53:02.000  2019-10-23T18:00:45.000   
9997         1  2019-10-23T17:07:16.000  2019-10-23T17:11:35.000   
9998         1  2019-10-23T17:38:26.000  2019-10-23T17:49:28.000   
9999         1  2019-10-23T17:22:14.000  2019-10-23T17:52:09.000   

      passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
9995                2            1.3           1                  N   
9996                1            1.4           1                  N   
9997                1            0.7           1                  N   
9998                2            2.5           1                  N   
9999                1            3.0           1                  N   

      pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
9995           238           239             1          8.0    3.5      0.5   
9996           239           166             2          8.0    3.5      0.5   
9997           166           152             2          5.0    1.0      0.5   
9998           151            42             1         10.0    1.0      0.5   
9999           237            68             1         19.0    3.5      0.5   

      tip_amount  tolls_amount  improvement_surcharge  total_amount  \
9995        2.46           0.0                    0.3         14.76   
9996        0.00           0.0                    0.3         12.30   
9997        0.00           0.0                    0.3          6.80   
9998        0.00           0.0                    0.3         11.80   
9999        2.50           0.0                    0.3         25.80   

      congestion_surcharge  
9995                   2.5  
9996                   2.5  
9997                   0.0  
9998                   0.0  
9999                   2.5

#### Exercise 1.2
##### Find the dimensions (number of rows and number of columns) in the data.

In [25]:
taxi.shape

(10000, 18)

#### Exercise 1.3
##### Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.

In [51]:
taxi[['fare_amount','tip_amount','tolls_amount','total_amount']].describe()

fare_amount    tip_amount  tolls_amount  total_amount
count  10000.000000  10000.000000  10000.000000  10000.000000
mean      15.106313      2.634494      0.623447     22.564659
std       13.954762      3.409800      6.437507     19.209255
min      -52.000000      0.000000     -6.120000    -65.920000
25%        7.000000      0.000000      0.000000     12.375000
50%       10.000000      2.000000      0.000000     16.300000
75%       16.000000      3.250000      0.000000     22.880000
max      176.000000     43.000000    612.000000    671.800000

#### Exercise 1.4
##### Isolate the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` for the longest trip by distance (`trip_distance`).

In [37]:
longest_trip = taxi['trip_distance'].idxmax()

In [38]:
longest_trip

8338

In [41]:
taxi['fare_amount'].iloc[longest_trip]

np.float64(176.0)

In [42]:
taxi['tip_amount'].iloc[longest_trip]

np.float64(18.29)

In [44]:
taxi['tolls_amount'].iloc[longest_trip]

np.float64(6.12)

In [45]:
taxi['total_amount'].iloc[longest_trip]

np.float64(201.21)

In [43]:
taxi.iloc[longest_trip]

vendorid                                       2
tpep_pickup_datetime     2019-10-23T16:50:53.000
tpep_dropoff_datetime    2019-10-24T15:32:55.000
passenger_count                                1
trip_distance                              38.11
ratecodeid                                     5
store_and_fwd_flag                             N
pulocationid                                 132
dolocationid                                 265
payment_type                                   1
fare_amount                                176.0
extra                                        0.0
mta_tax                                      0.5
tip_amount                                 18.29
tolls_amount                                6.12
improvement_surcharge                        0.3
total_amount                              201.21
congestion_surcharge                         0.0
Name: 8338, dtype: object

In [54]:
taxi.loc[taxi.trip_distance.idxmax(),
    ['fare_amount','tip_amount','tolls_amount','total_amount']]

fare_amount      176.0
tip_amount       18.29
tolls_amount      6.12
total_amount    201.21
Name: 8338, dtype: object

---

### Section 2

#### Exercise 2.1
##### Read in the meteorite data from the `Meteorite_Landings.csv` file, rename the `mass (g)` column to `mass`, and drop all the latitude and longitude columns. Sort the result by mass in descending order.

In [84]:
Meteorite = pd.read_csv('../data/Meteorite_Landings.csv')

In [85]:
Meteorite = Meteorite.rename(columns = {'mass (g)':'mass'})

In [86]:
Meteorite.columns

Index(['name', 'id', 'nametype', 'recclass', 'mass', 'fall', 'year', 'reclat',
       'reclong', 'GeoLocation'],
      dtype='object')

In [87]:
Meteorite = Meteorite\
    .drop(columns=Meteorite.columns[-3:])\
    .sort_values ('mass', ascending=False)

In [88]:
Meteorite.head()

name     id nametype      recclass        mass   fall  \
16392             Hoba  11890    Valid     Iron, IVB  60000000.0  Found   
5373         Cape York   5262    Valid   Iron, IIIAB  58200000.0  Found   
5365   Campo del Cielo   5247    Valid  Iron, IAB-MG  50000000.0  Found   
5370     Canyon Diablo   5257    Valid  Iron, IAB-MG  30000000.0  Found   
3455           Armanty   2335    Valid    Iron, IIIE  28000000.0  Found   

                         year  
16392  01/01/1920 12:00:00 AM  
5373   01/01/1818 12:00:00 AM  
5365   12/22/1575 12:00:00 AM  
5370   01/01/1891 12:00:00 AM  
3455   01/01/1898 12:00:00 AM

#### Exercise 2.2
##### Using the meteorite data from the `Meteorite_Landings.csv` file, update the `year` column to only contain the year, convert it to a numeric data type, and create a new column indicating whether the meteorite was observed falling before 1970. Set the index to the `id` column and extract all the rows with IDs between 10,036 and 10,040 (inclusive) with `loc[]`.

###### **Hint 1**: Use `year.str.slice()` to grab a substring.

###### **Hint 2**: Make sure to sort the index before using `loc[]` to select the range.

###### **Bonus**: There's a data entry error in the `year` column. Can you find it? (Don't spend too much time on this.)

In [91]:
Meteorite = Meteorite.assign(
    year = lambda x: pd.to_numeric(x.year.str.slice(6,10)),
    pre_1970 = lambda x: (x.fall == 'Fell') &(x.year <1970)
).set_index ('id')

In [92]:
Meteorite.sort_index().loc[10_036:10_040]

name nametype         recclass      mass   fall    year  pre_1970
id                                                                           
10036     Enigma    Valid               H4      94.0  Found  1967.0     False
10037       Enon    Valid  Iron, ungrouped     763.0  Found  1883.0     False
10038      Enshi    Valid               H5    8000.0   Fell  1974.0     False
10039  Ensisheim    Valid              LL6  127000.0   Fell  1491.0      True

In [83]:
Meteorite.head()

16392    01/01/1920
5373     01/01/1818
5365     12/22/1575
5370     01/01/1891
3455     01/01/1898
Name: year, dtype: object

In [93]:
Meteorite.year.describe()

count    45425.000000
mean      1991.828817
std         25.052766
min        860.000000
25%       1987.000000
50%       1998.000000
75%       2003.000000
max       2101.000000
Name: year, dtype: float64

In [95]:
Meteorite.query(f'year>{pd.Timestamp('today').year}')

name nametype recclass  mass   fall    year  pre_1970
id                                                                           
57150  Northwest Africa 7701    Valid      CK6  55.0  Found  2101.0     False

#### Exercise 2.3
##### Using the meteorite data from the `Meteorite_Landings.csv` file, create a pivot table that shows both the number of meteorites and the 95th percentile of meteorite mass for those that were found versus observed falling per year from 2005 through 2009 (inclusive). Hint: Be sure to convert the `year` column to a number as we did in the previous exercise.

#### Exercise 2.4
##### Using the meteorite data from the `Meteorite_Landings.csv` file, compare summary statistics of the mass column for the meteorites that were found versus observed falling.

#### Exercise 2.5
##### Using the taxi trip data in the `2019_Yellow_Taxi_Trip_Data.csv` file, resample the data to an hourly frequency based on the dropoff time. Calculate the total `trip_distance`, `fare_amount`, `tolls_amount`, and `tip_amount`, then find the 5 hours with the most tips.

--- 

### Section 3

#### Exercise 3.1
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create box plots for traveler throughput for each year in the data. Hint: Pass `kind='box'` into the `plot()` method to generate box plots.

#### Exercise 3.2
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create a heatmap that shows the 2019 TSA median traveler throughput by day of week and month.

#### Exercise 3.3
##### Annotate the medians in the box plot from *[Exercise 3.1](#Exercise-3.1)*. Hint: The `x` coordinates will be 1, 2, and 3 for 2019, 2020, and 2021, respectively. Alternatively, to avoid hardcoding values, you can use the `Axes.get_xticklabels()` method, in which case you should look at the [documentation](https://matplotlib.org/stable/api/text_api.html) for the `Text` class.